harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,519 @@
+
+
+class _Disk(_MetricGroupCollector):
+    group = "disk"
+
+    _disk_paths = ["/", constants.SHARED_FS_CONTAINER_PATH]
+
+    def __init__(self) -> None:
+        # Set initial values for throughput calculations.
+        self._interval_start_ts = time.time()
+        self._interval_start_vals = psutil.disk_io_counters()
+
+        # Initialize accessible disk paths.
+        self._paths = []
+        for path in self._disk_paths:
+            try:
+                psutil.disk_usage(path)
+                self._paths.append(path)
+            except Exception:
+                pass
+
+        super().__init__()
+
+    def sample_metrics(self) -> None:
+        ts = time.time()
+        vals = psutil.disk_io_counters()
+
+        read_thru = (vals.read_bytes - self._interval_start_vals.read_bytes) / (
+            ts - self._interval_start_ts
+        )
+        write_thru = (vals.write_bytes - self._interval_start_vals.write_bytes) / (
+            ts - self._interval_start_ts
+        )
+        iops = (vals.read_count + vals.write_count) - (
+            self._interval_start_vals.read_count + self._interval_start_vals.write_count
+        ) / (ts - self._interval_start_ts)
+        self._interval_start_ts, self._interval_start_vals = ts, vals
+
+        metrics = {
+            "disk_iops": iops,
+            "disk_throughput_read": read_thru,
+            "disk_throughput_write": write_thru,
+        }
+
+        for path in self._paths:
+            disk_usage = psutil.disk_usage(path)
+            metrics.update({path: {"disk_util": disk_usage.percent}})
+        self.metric_samples.append(metrics)
+
+
+class _Memory(_MetricGroupCollector):
+    group = "memory"
+
+    def sample_metrics(self) -> None:
+        free_mem_bytes = psutil.virtual_memory().available
+        metrics = {
+            "memory_free": free_mem_bytes / 1e9,
+        }
+        self.metric_samples.append(metrics)
+
+
+class _CPU(_MetricGroupCollector):
+    group = "cpu"
+
+    def sample_metrics(self) -> None:
+        cpu_util = psutil.cpu_percent()
+        metrics = {
+            "cpu_util_simple": cpu_util,
+        }
+        self.metric_samples.append(metrics)
+
+
+class _GPU(_MetricGroupCollector):
+    group = "gpu"
+
+    def __init__(self) -> None:
+        super().__init__()
+
+        self._pynvml_device_handles: Dict[str, Any] = {}
+
+        if pynvml:
+            self._init_pynvml()
+        else:
+            logging.warning("pynvml module not found. GPU metrics will not be collected.")
+
+    def _init_pynvml(self) -> None:
+        """Initialize the pynvml library and validate methods.
+
+        Call the NVML library methods we'll be using to validate that they are accessible.
+        Sometimes NVML initializes successfully but individual device methods will fail.
+
+        If any NVML method fails for any GPU device, no GPU metrics will be collected for
+        all GPU devices.
+        """
+        try:
+            assert pynvml
```

Comment:
nit: could you please pull this outside the try? Makes my IDE happier for the except block.